src/engine/strat_engine/backstore/backstore.rs

```diff
@@ -146,46 +165,247 @@ impl DataTier {
+        let table = map_to_dm(&coalesced);
+
+        cache_device.set_cache_table(dm, table)?;
+        cache_device.resume(dm)?;
+
+        self.cache_segments = coalesced;
+
+        Ok(uuids)
+    }
+
+    /// Setup a new CacheTier struct from the block_mgr.
+    ///
+    /// Returns the CacheTier and the cache device that was created.
+    ///
+    /// WARNING: metadata changing event
+    pub fn new(dm: &DM,
+               mut block_mgr: BlockDevMgr,
+               origin: LinearDev)
+               -> EngineResult<(CacheTier, CacheDev)> {
+        let avail_space = block_mgr.avail_space();
+        // TODO: check whether this should be increased
+        let meta_space = Sectors(4 * IEC::Ki);
+
+        assert!(meta_space < avail_space,
+                "every block device must be at least one GiB");
+
+        let mut segments = block_mgr
+            .alloc_space(&[meta_space, avail_space - meta_space])
+            .expect("asked for exactly the space available, must get");
+
+        let cache_segments = segments.pop().expect("segments.len() == 2");
+        let meta_segments = segments.pop().expect("segments.len() == 1");
+
+        let (dm_name, dm_uuid) = format_backstore_ids(block_mgr.pool_uuid(), CacheRole::MetaSub);
+        let meta = LinearDev::setup(dm, &dm_name, Some(&dm_uuid), map_to_dm(&meta_segments))?;
+
+        // See comment in ThinPool::new() method
+        wipe_sectors(&meta.devnode(), Sectors(0), meta.size())?;
+
+
+        let (dm_name, dm_uuid) = format_backstore_ids(block_mgr.pool_uuid(), CacheRole::CacheSub);
+        let cache = LinearDev::setup(dm, &dm_name, Some(&dm_uuid), map_to_dm(&cache_segments))?;
+
+        let (dm_name, dm_uuid) = format_backstore_ids(block_mgr.pool_uuid(), CacheRole::Cache);
+        let cd = CacheDev::new(dm,
+                               &dm_name,
+                               Some(&dm_uuid),
+                               meta,
+                               cache,
+                               origin,
+                               MIN_CACHE_BLOCK_SIZE)?;
+
+        Ok((CacheTier {
+                block_mgr,
+                meta_segments,
+                cache_segments,
+            },
+            cd))
+    }
 }
 
+
+/// This structure can allocate additional space to the upper layer, but it
+/// cannot accept returned space. When it is extended to be able to accept
+/// returned space the allocation algorithm will have to be revised.
+///
+/// self.linear.is_some() XOR self.cache.is_some()
+/// self.cache.is_some() <=> self.cache_tier.is_some()
 #[derive(Debug)]
 pub struct Backstore {
+    /// A cache DM Device.
+    cache: Option<CacheDev>,
+    /// Coordinate handling of blockdevs that back the cache. Optional, since
+    /// this structure can operate without a cache.
+    cache_tier: Option<CacheTier>,
+    /// Coordinates handling of the blockdevs that form the base.
     data_tier: DataTier,
+    /// A linear DM device.
+    linear: Option<LinearDev>,
```

Comment:
Do to sharing issues, this is the only way I can allow for the possibilty of EITHER cache OR linear. If the cache exists, then the linear device has to end up owned by the cache device. That operation requires a fair amount of trickiness, see ```add_blockdevs```. The ```either``` crate does not help, because objects of either type are not mutable. The header comment asserts the EITHER/OR invariant, which is enforced wherever relevant, and also in the test code.